### PR TITLE
Threescale 1918: Add type checking and tests to header

### PR DESCRIPTION
**Pull request by damianpm**
**What this PR does / why we need it**:

Add Flow type checking and JS tests to the PF4/React Header, added in #618 

**Which issue(s) this PR fixes** 

https://issues.jboss.org/browse/THREESCALE-1918

**Verification steps** 

Pull branch
`npm run jest`
`npx flow`

### CODE REVIEW

File: app/javascript/src/Navigation/components/header/AccountSettingsMenu.jsx [hunk 1]
```diff
@@ -1,6 +1,9 @@
-import React from 'react'
+// @flow
 
-const AccountSettingsMenu = ({accountSettingsLink, accountSettingsClass}) => (
+import * as React from 'react'
+import { ToolbarProps } from 'Navigation/components/header/types'
+
+const AccountSettingsMenu = ({accountSettingsLink, accountSettingsClass}: ToolbarProps): React.Node => (
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/AccountSettingsMenu.jsx]: Returning type is not required here, it's inferred
```suggestion
const AccountSettingsMenu = ({accountSettingsLink, accountSettingsClass}: ToolbarProps) => (
```

Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/AccountSettingsMenu.jsx]: On the other hand, `ToolbarProps` looks wrong here. Since the component is `AccountSettingsMenu`, its props type should be `AccountSettingsMenuProps`. Does that make sense? We would have a conflict with `Toolbar`, since it has as props `{ toolbarProps, docsProps }` so I think `toolbarProps` are kinda wrong there and should be renamed.

File: spec/javascripts/Navigation/header/header.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    title: 'Dashboard',
+    target: '_self',
+    className: `Header-link`
+  }
+}
+
+describe('<Header/>', () => {
+  it('renders <Header/> component', () => {
+    const wrapper = shallow(<Header {...Props} />)
+    expect(wrapper.props()).toHaveProperty('logo')
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/header.spec.jsx]: You are asserting here properties, not that it's rendered.

File: spec/javascripts/Navigation/header/header.spec.jsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    expect(wrapper.props()).toHaveProperty('logoProps')
+    expect(wrapper.props()).toHaveProperty('toolbar')
+    expect(wrapper.props()).toHaveProperty('avatar')
+    expect(wrapper.find(PageHeader)).toHaveLength(1)
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/header.spec.jsx]: This is a different assert, it expects to have a `PageHeader`.

File: spec/javascripts/Navigation/header/SessionMenu.spec.jsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  avatarLinkClass: 'my-class',
+  impersonated: 'true',
+  accountName: 'account-name',
+  displayName: 'display-name',
+  logoutPath: 'logout',
+  username: 'username'
+}
+
+describe('<SessionMenu/>', () => {
+  it('renders <SessionMenu/> component when impersonated', () => {
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/SessionMenu.spec.jsx]: Specs should describe the functionality of a component. IMO this is not very clear here. Instead of writing a bunch of repetitive and non-flexible assertions, what about writing smaller `it` blocks that describe small pieces of functionality?

For instance:
- If a `.PopNavigation--session` className is required and not just a styling decision, then assert `it should have a X class`
- If there should be more than 1 `.fa-bolt` (which is not clear what this is) then something like `it should have 2 X`
- You could test sign out button in another describe block, with multiple assertions: is it there? what happens when clicked? etc

File: spec/javascripts/Navigation/header/AccountSettingsMenu.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import { AccountSettingsMenu } from 'Navigation/components/header/AccountSettingsMenu'
+
+Enzyme.configure({ adapter: new Adapter() })
+
+const props = {
+  accountSettingsLink: 'my-link',
+  accountSettingsClass: 'my-class'
+}
+
+describe('<AccountSettingsMenu/>', () => {
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/AccountSettingsMenu.spec.jsx]: From a presentational point of view it's a bit redundant to `describe` the file being tested:
```
 PASS  spec/javascripts/Navigation/header/AccountSettingsMenu.spec.jsx
  <AccountSettingsMenu/>
    ✓ renders <AccountSettingsMenu/> component (18ms)
```
Usually describe is used to separate blocks of functionality. For instance `AccountSettingsMenu` could be open or close, then:
```js
describe('when dropdown is displayed', () => {
  it('should hide the dropdown when clicked')
}
describe('when dropdown is hidden', () => {
  it('should display a dropdown when clicked')
}
```
Otherwise it's enough with simply using `it`.

File: spec/javascripts/Navigation/header/Brand.spec.jsx [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+const props = {
+  accountSettingsLink: 'my-link',
+  accountSettingsClass: 'my-class'
+}
+
+describe('<Brand/>', () => {
+  it('renders <Brand/> component', () => {
+    const wrapper = shallow(<Brand {...props} />)
+    expect(wrapper.find('.u-header-brand')).toHaveLength(1)
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/Brand.spec.jsx]: Use `exists()` to assert the presence o an element and `toHaveLenght` to assert the length of a collection of children.

File: spec/javascripts/Navigation/header/DocumentationMenu.spec.jsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    expect(wrapper.find('ul.PopNavigation-list')).toHaveLength(1)
+    expect(wrapper.find(DocumentationMenuItem)).toHaveLength(4)
+  })
+
+  it('renders <DocumentationMenu/> component when no Saas', () => {
+    let props = Object.assign({}, Props, {isSaas: null})
+    const wrapper = shallow(<DocumentationMenu {...props} />)
+    expect(wrapper.find('.PopNavigation--docs')).toHaveLength(1)
+    expect(wrapper.find('ul.PopNavigation-list')).toHaveLength(1)
+    expect(wrapper.find(DocumentationMenuItem)).toHaveLength(3)
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/DocumentationMenu.spec.jsx]: Perhaps this `3` should be extracted as a constant or described somehow.

File: app/javascript/src/Navigation/components/header/DocumentationMenu.jsx [hunk 8]
```diff
@@ -1,11 +1,14 @@
-import React from 'react'
+// @flow
+
+import * as React from 'react'
 import { DocumentationMenuItem } from 'Navigation/components/header'
+import { DocsProps } from 'Navigation/components/header/types'
 
 const renderMenuItems = (docsLinksClass, items) => items.map(
```
Comment (didierofrivia) [inline app/javascript/src/Navigation/components/header/DocumentationMenu.jsx]: Types for `docsLinksClass` and `items` ?

File: app/javascript/src/Navigation/components/header/DocumentationMenu.jsx [hunk 9]
```diff
@@ -1,11 +1,14 @@
...
   item => <DocumentationMenuItem key={item.text.replace(' ', '')} docsLinksClass={docsLinksClass} item={item}/>
 )
 
-const DocumentationMenu = ({docsLink, isSaas, docsLinksClass, customerPortalLink, apiDocsLink, liquidReferenceLink, whatIsNewLink}) => {
+const DocumentationMenu = ({docsLink, isSaas, docsLinksClass, customerPortalLink, apiDocsLink, liquidReferenceLink, whatIsNewLink}: DocsProps): React.Node => {
```
Comment (didierofrivia) [inline app/javascript/src/Navigation/components/header/DocumentationMenu.jsx]: Maybe here having them in different lines for easier read, then adding types?

Comment (didierofrivia) [reply app/javascript/src/Navigation/components/header/DocumentationMenu.jsx]: Forget this ^^

File: app/javascript/src/Navigation/components/header/DocumentationMenuItem.jsx [hunk 10]
```diff
@@ -1,6 +1,15 @@
-import React from 'react'
+// @flow
 
-const DocumentationMenuItem = ({docsLinksClass, item}) => (
+import * as React from 'react'
+
+type ItemType = {
```
Comment (didierofrivia) [inline app/javascript/src/Navigation/components/header/DocumentationMenuItem.jsx]: I prefer this way over importing types from another file when it's just used in one component. That way you can see at glance what is going on.

File: spec/javascripts/Navigation/header/DocumentationMenu.spec.jsx (hunk 7 above)
Comment (didierofrivia) [reply spec/javascripts/Navigation/header/DocumentationMenu.spec.jsx]: Yeah, I understand there's an extra Item for SaaS, but would love to have it reflected in the test.

File: spec/javascripts/Navigation/header/SessionMenu.spec.jsx (hunk 4 above)
Comment (didierofrivia) [reply spec/javascripts/Navigation/header/SessionMenu.spec.jsx]: I agree with this. When a component is built up with many pieces, those needs to be described and asserted in separated `it` blocks. It needs to be in this case very descriptive, since apparently there's a lot of pieces that might or not vary depending on props.

**Comment by damianpm** [comment]
@didierofrivia @josemigallas PR updated

File: app/javascript/src/Navigation/components/header/AccountSettingsMenu.jsx [hunk 11]
```diff
@@ -1,6 +1,9 @@
-import React from 'react'
+// @flow
 
-const AccountSettingsMenu = ({accountSettingsLink, accountSettingsClass}) => (
+import * as React from 'react'
+import { AccountSettingsProps } from 'Navigation/components/header/types'
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/AccountSettingsMenu.jsx]: I wouldn't put Prop and State type definitions in other files, even if it were used elsewhere but it helps encapsulation.

File: app/javascript/src/Navigation/components/header/Brand.jsx [hunk 12]
```diff
@@ -1,4 +1,5 @@
-import React from 'react'
+// @flow
+import * as React from 'react'
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/Brand.jsx]: Not really important but I think there's no need to import all. This is done in Flow only to be able to use React types.

File: app/javascript/src/Navigation/components/header/DocumentationMenu.jsx [hunk 13]
```diff
@@ -1,11 +1,21 @@
-import React from 'react'
+// @flow
+
+import * as React from 'react'
 import { DocumentationMenuItem } from 'Navigation/components/header'
+import { DocsProps } from 'Navigation/components/header/types'
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/DocumentationMenu.jsx]: Same as comment above, also I think the proper name is `DocumentationMenuProps` or simply `Props` since it's private.

File: app/javascript/src/Navigation/components/header/DocumentationMenu.jsx [hunk 14]
```diff
@@ -1,11 +1,21 @@
...
+
+type ItemType = {
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/DocumentationMenu.jsx]: This is a bit confusing and also redundant, since it's the same type definition as in DocumentationMenuItem. I think either we import this type from that class or we remove `renderMenuItems`. (My preference for the latter)

File: app/javascript/src/Navigation/components/header/Header.jsx [hunk 15]
```diff
@@ -1,12 +1,17 @@
...
+// @flow
+
+import * as React from 'react'
 import { PageHeader } from '@patternfly/react-core'
 import { Brand, SessionMenu, Toolbar } from 'Navigation/components/header'
+import { LogoProps, AccountSettingsProps, DocsProps, AvatarProps } from 'Navigation/components/header/types'
 
-const Header = ({logoProps, toolbarProps, docsProps, avatarProps}) => (
+const Header = ({logoProps, accountSettingsProps, docsProps, avatarProps}: {
+  logoProps: LogoProps, accountSettingsProps: AccountSettingsProps, docsProps: DocsProps, avatarProps: AvatarProps
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/Header.jsx]: It would be clearer to instead declare a `type Props` for this whole line.

File: app/javascript/src/Navigation/components/header/SessionMenu.jsx [hunk 16]
```diff
@@ -1,8 +1,11 @@
-import React from 'react'
+// @flow
+
+import * as React from 'react'
 import { Avatar } from '@patternfly/react-core'
 import avatarImg from 'Navigation/images/img_avatar.svg'
+import { AvatarProps } from 'Navigation/components/header/types'
 
-const SessionMenu = ({avatarLinkClass, impersonated, accountName, displayName, logoutPath, username}) => (
+const SessionMenu = ({avatarLinkClass, impersonated, accountName, displayName, logoutPath, username}: AvatarProps) => (
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/SessionMenu.jsx]: `SessionMenuProps` would be better IMO or simply `Props` if we declare them inside this file. If needed somewhere else like Header.jsx it can be imported with a different name. I guess we fall again into personal preferences but this is what I'd do:
```js
// SessionMenu.jsx
export type Props = {
  avatarLinkClass: ...,
  impersonated: ...,
  ...
}

// Header.jsx
import type { Props as SessionMenuProps } from 'components/SessionMenu'
```

File: app/javascript/src/Navigation/components/header/Toolbar.jsx [hunk 17]
```diff
@@ -1,11 +1,14 @@
-import React from 'react'
+// @flow
+
+import * as React from 'react'
 import { Toolbar as PFToolbar, ToolbarGroup, ToolbarItem } from '@patternfly/react-core'
 import { AccountSettingsMenu, DocumentationMenu } from 'Navigation/components/header/'
+import { AccountSettingsProps, DocsProps } from 'Navigation/components/header/types'
 
-const Toolbar = ({ toolbarProps, docsProps }) => <PFToolbar>
+const Toolbar = ({ accountSettingsProps, docsProps }: { accountSettingsProps: AccountSettingsProps, docsProps: DocsProps }) => <PFToolbar>
```
Comment (josemigallas) [inline app/javascript/src/Navigation/components/header/Toolbar.jsx]:
```suggestion
const Toolbar = ({ accountSettingsProps, docsProps }: { accountSettingsProps: AccountSettingsProps, docsProps: DocsProps }) => (
  <PFToolbar>
```
😁

File: jest.config.js [hunk 18]
```diff
@@ -3,7 +3,8 @@ module.exports = {
   moduleNameMapper: {
     'c3': '<rootDir>/__mocks__/c3.js',
     '\\.(css|less|sass|scss)$': '<rootDir>/spec/javascripts/__mocks__/styleMock.js',
-    '\\.(gif|ttf|eot|svg)$': '<rootDir>/spec/javascripts/__mocks__/fileMock.js'
+    '\\.(gif|ttf|eot)$': '<rootDir>/spec/javascripts/__mocks__/fileMock.js',
+    '\\.(svg)$': '<rootDir>/spec/javascripts/__mocks__/empty-module.js'
```
Comment (josemigallas) [inline jest.config.js]: Why this change?

File: spec/javascripts/Navigation/header/SessionMenu.spec.jsx [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+  it('should display proper message and not any `fa-bolt` icon when not impersonated', () => {
+    const unImpersonatedProps = Object.assign({}, props, { impersonated: null })
+    const wrapper = shallow(<SessionMenu {...unImpersonatedProps} />)
+    expect(wrapper.find('.fa-bolt')).toHaveLength(0)
+    expect(wrapper.find('.PopNavigation-info').text()).not.toContain('Impersonating a virtual admin user from')
+    expect(wrapper.find('.PopNavigation-info').text()).toContain('Signed in to')
+  })
+  it('should render one <Avatar/> component', () => {
+    const wrapper = shallow(<SessionMenu {...props} />)
+    expect(wrapper.exists(Avatar)).toEqual(true)
```
Comment (josemigallas) [inline spec/javascripts/Navigation/header/SessionMenu.spec.jsx]:
```suggestion
    expect(wrapper.find(Avatar)).toHaveLength(1)
```

File: app/javascript/src/Navigation/components/header/SessionMenu.jsx (hunk 16 above)
Comment (didierofrivia) [reply app/javascript/src/Navigation/components/header/SessionMenu.jsx]: I'd follow as a rule just `Props` when resides on the same file, and the type definition is just above the fn definition.

Comment (didierofrivia) [reply app/javascript/src/Navigation/components/header/SessionMenu.jsx]: When it's imported, then same name of the component + "Props" as in `ComponentProps`

Comment (didierofrivia) [reply app/javascript/src/Navigation/components/header/SessionMenu.jsx]: When it's imported, then same name of the component + "Props" as in `ComponentProps`

**Comment by didierofrivia** [comment]
I agree with @josemi only partially because in this case, when creating a UI component that specifically was design a certain way, the structure of the html is important to assert. Also, there's no shame in being repetitive in a test, you don't need to go meta programming when writing it, that's the ultimate anti pattern. On the other hand, if there's no relevant information, behaviour or design, feel free to only test it was mounted or not tested at all. The test should be meaningful in the sense it needs to tell the story of what a component it supposed to look (if by design it's important) but mostly how it supposed to behave.
